Add support for running process asynchronously

JoshuaWatt opened the pull request:
<!---
Describe what your pull request does,
i.e. fix this bug and how, add a feature, fix documentation…
If you add a feature, please answer these questions:
- what do you use the feature for?
- how does labgrid benefit as a testing library from the feature?
- how did you verify the feature works?
- if hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?
--->
**Description**
Adds support for running processes asynchronously with the main test code. Attempts to incorporate the feed back from #552 

<!---
This checklist roughly outlines the steps for new features, remove and add tasks as needed:
--->
**Checklist**
- [ ] Documentation for the feature
- [x] Tests for the feature 
<!---
If you add a driver/resource or modifiy one:
--->
- [ ] The arguments and description in doc/configuration.rst have been updated
<!---
If you add a feature other drivers/resources can benefit from:
--->
- [ ] Add a section on how to use the feature to doc/usage.rst
<!---
A library feature which other developers can use:
--->
- [ ] Add a section on how to use the feature to doc/development.rst
<!---
Provide a short summary for the CHANGES.rst file
--->
- [ ] CHANGES.rst has been updated
<!---
Did you test the change locally? If yes, best to mention how you did it in the description section.
--->
- [x] PR has been tested
<!---
If your PR touched the man pages they have to be regenerated by calling make in the man subdirectory of the project
--->
- [ ] Man pages have been regenerated

<!---
In case your PR fixes a bug, please reference it in the next line, i.e.
Fixes #[insert number without brackets here]
--->

JoshuaWatt committed:
commandprotocol: Add Process Protocol

Extends the command protocol to support "background" processes.

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
marker: Add ConsoleMarkerProcess helper class

Many different drivers need to implement "marker" style command
processes where the command output is terminated with a random marker
then the command status code and a prompt. Instead of making each driver
implement this logic over and over again, implement a common class that
handles this type of command output.

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
bareboxdriver: Implement start_process()

Implements the start_process() API for "background" processes using the
ConsoleMarkerProcess helper class. Rework the _run() API to use
a background process behind the scenes

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
shelldriver: Implement start_process()

Implements the start_process() API for "background" processes using the
ConsoleMarkerProcess helper class. Rework the _run() API to use
a background process behind the scenes

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>
[rouven.czerwinski@linaro.org: adjusted to UBoot timestamp removal]
Signed-off-by: Rouven Czerwinski <rouven.czerwinski@linaro.org>

JoshuaWatt committed:
ubootdriver: Implement start_process()

Implements the start_process() API for "background" processes using the
ConsoleMarkerProcess helper class. Rework the _run() API to use
a background process behind the scenes

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
sshdriver: Implement start_process()

Implements the start_process() API for SSH connections. The
SSHDriverProcess is a wrapper around pexpect.spawn()

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
timeout: Allow integer timeouts

Allowing the class to take an integer for a timeout instead of a strict
float seems more convenient to users than forcing it at a higher level,
or breaking usage by changing other APIs to require floats (i.e.
run_check())

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

Emantor commented:
Overall this design looks like what we had in mind in #552. Thank you very much for the implementation, looking forward to testing this on my lab at home.

JoshuaWatt commented:
It is unfortunately a little hard to write automated tests for most of this since it requires actual hardware interaction, but I'll take a stab at it. I was able to test on actual hardware (Raspberry Pi 4 w/ u-boot) with the following test suite:
```python
import socket
import os
import time

def test_uboot_async(uboot_command):
    with uboot_command.popen("echo hello") as p:
        assert p.read(4).decode("utf-8").splitlines() == ["hell"]
        assert p.read().decode("utf-8").splitlines() == ["o"]

    with uboot_command.popen("echo bzap hello bzap") as p:
        index, before, match, after = p.expect("hello")
        assert index == 0
        assert match.group(0).decode("utf-8") == "hello"
        assert before.decode("utf-8") == "bzap "
        assert after.decode("utf-8") == "hello"
        assert p.read().decode("utf-8").splitlines() == [" bzap"]

def test_ssh_async(ssh_command):
    with ssh_command.popen("echo hello") as p:
        assert p.read(4).decode("utf-8").splitlines() == ["hell"]
        assert p.read().decode("utf-8").splitlines() == ["o"]

    with ssh_command.popen("echo bzap hello bzap") as p:
        index, before, match, after = p.expect("hello")
        assert index == 0
        assert match.group(0).decode("utf-8") == "hello"
        assert before.decode("utf-8") == "bzap "
        assert after.decode("utf-8") == "hello"
        assert p.read().decode("utf-8").splitlines() == [" bzap"]

def test_shell_nc(shell_command, ssh_command):
    with shell_command.popen("nc -l -p 5555") as p:
        # Wait for netcat to start listening
        time.sleep(2)
        with ssh_command.forward_local_port(5555) as local_port:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.connect(("localhost", local_port))
                s.sendall(b"Hello World")
            p.expect(b"Hello World")
            time.sleep(2)

def test_ssh_nc(ssh_command):
    with ssh_command.popen("nc -l -p 5555 127.0.0.1") as p:
        # Wait for netcat to start listening
        time.sleep(2)
        with ssh_command.forward_local_port(5555) as local_port:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.connect(("localhost", local_port))
                s.sendall(b"Hello World")
            p.expect(b"Hello World")
            time.sleep(2)


def test_shell_async(shell_command):
    with shell_command.popen("echo hello") as p:
        assert p.read(4).decode("utf-8").splitlines() == ["hell"]
        assert p.read().decode("utf-8").splitlines() == ["o"]

    with shell_command.popen("echo bzap hello bzap") as p:
        index, before, match, after = p.expect("hello")
        assert index == 0
        assert match.group(0).decode("utf-8") == "hello"
        assert before.decode("utf-8") == "bzap "
        assert after.decode("utf-8") == "hello"
        assert p.read().decode("utf-8").splitlines() == [" bzap"]

def test_shell_interrupt(shell_command):
    with shell_command.popen("yes") as p:
        l = p.read(1024).decode("utf-8").splitlines()
        assert l == ["y"] * len(l)

    shell_command.run_check("true")
```

Note that I wasn't able to test the barebox driver, so manual testing there would be appreciated

JoshuaWatt commented:
I added some tests, which revealed a few bugs that needed fixed. I think it's all good to go now though (although I still have no way to test barebox)

Bastian-Krause commented on labgrid/driver/shelldriver.py:
```diff
@@ -9,13 +9,13 @@
 
 import attr
-from pexpect import TIMEOUT
+from pexpect import TIMEOUT, EOF
```
`EOF` is unused.

Bastian-Krause commented on labgrid/driver/sshdriver.py:
```diff
@@ -8,18 +8,88 @@
 import tempfile
 import time
+import pexpect
+import shlex
```
`shlex` is an import from the Python Standard Library, so move it up (standard imports first, then third party, then local package)

Bastian-Krause commented on labgrid/protocol/commandprotocol.py:
```diff
@@ -1,6 +1,85 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def read_nonblocking(size=1, timeout=-1):
```
If this is a method, it misses the `self` argument.

Bastian-Krause commented on labgrid/driver/ubootdriver.py:
```diff
@@ -4,11 +4,11 @@
 
 import attr
-from pexpect import TIMEOUT
+from pexpect import TIMEOUT, EOF
```
`EOF` is unused.

Bastian-Krause commented on labgrid/util/expect.py:
```diff
@@ -33,3 +35,12 @@ def read_nonblocking(self, size=1, timeout=-1):
+            self.send(bytes([ord_]))
+        except ValueError:
+            raise NotImplementedError("Sending control character {} is not supported yet".format(char))
+
```
Unnecessary trailing newline.

Bastian-Krause commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,178 @@
 import random
 import string
+import re
+import time
```
Unused import.

Bastian-Krause commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,178 @@
+
+from ..protocol import CommandProcessProtocol
+from ..step import step
+from ..util import Timeout, OptionalTimeout
```
`Timeout` is unused.

JoshuaWatt replied:
Done

Bastian-Krause commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,178 @@
+from ..protocol import CommandProcessProtocol
+from ..step import step
+from ..util import Timeout, OptionalTimeout
+from pexpect import EOF, TIMEOUT, spawn
```
This is a third party import. Move it up, so it is above the local package imports (standard imports first, then third party, then local package).

JoshuaWatt replied:
Done

Bastian-Krause reviewed (dismissed):
Great work! I found some trivial things. I'll have to dive deeper into this later.

This should be mentioned in `CHANGES.rst` and probably needs some documentation elsewhere?

JoshuaWatt committed:
Update documentation for asynchronous processes

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt commented:
> Great work! I found some trivial things. I'll have to dive deeper into this later.
> 
> This should be mentioned in `CHANGES.rst` and probably needs some documentation elsewhere?

I added a little to `CHANGES.rst`; I can add more documentation but I'm not really sure where it should go.

JoshuaWatt commented on labgrid/protocol/commandprotocol.py:
```diff
@@ -1,6 +1,85 @@
+        A timeout of None means to wait forever for either data or EOF
+
+        Note that the timer is per-character, not for the entire buffer
+        """
```
Because I forgot to add it :)

Bastian-Krause commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,178 @@
+
+    @step(args=['size', 'timeout'], result=True)
+    def read(self, size=-1, *, timeout=None):
+        t = OptionalTimeout(timeout, self._console.timeout)
```
`QEMUDriver` and `ExternalConsoleDriver` do not implement a timeout attribute.

Bastian-Krause commented on labgrid/driver/bareboxdriver.py:
```diff
@@ -79,25 +91,21 @@ def _run(self, cmd: str, *, timeout: int = 30, codec: str = "utf-8", decodeerror
+                output = p.read(timeout=timeout)
+                # Remove VT100 Codes and split by newline
+                data = self.re_vt100.sub('', output.decode('utf-8')).split('\r\n')[1:-1]
+                self.logger.debug(f"Received Data: {data}")
```
log functions should not use f-strings, but rather lazy % formatting. There are multiple occurrences of this spread in this PR.

Bastian-Krause replied:
Yes, there is a reason. Here is a short wrap-up: https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

TL;DR: formatting is only done if required (i.e. the log level is reached) and errors are handled gracefully.

JoshuaWatt replied:
Reading the article has confused me more; IIUC, the article is suggesting that logging should use lazy evaluation by passing format arguments as parameters (_not_ using `%`), in which case the log message would be something like `self.logger.debug("Received data: %s", data)`. However, it seems you are suggesting I use `%` instead, which is _not_ a lazy evalutation and therefore not superior (at least in terms of what the article is discussing) from f-strings.

I _think_ what you want to do is either use actual lazy evaluation for log messages, or non-lazy f-strings for consistency with everywhere else f-strings are used

Bastian-Krause replied:
With "% formatting" I mean `logger.debug("%s", data)`.

JoshuaWatt replied:
OK, I will do that

Bastian-Krause commented on CHANGES.rst:
```diff
@@ -20,6 +20,8 @@ New Features in 0.4.0
 - AndroidFastbootDriver now supports booting/flashing images preconfigured in
   the environment configuration.
+- CommandProtocol now supports a popen() method to run a process on the target
+  asynchronously with the pytest code.
```
Why mention pytest here? The PR does not touch anything pytest-specific.

Bastian-Krause commented on labgrid/driver/ubootdriver.py:
```diff
@@ -69,26 +69,25 @@ def on_deactivate(self):
+                data = self.re_vt100.sub(
+                    '', output.decode('utf-8'), count=1000000
+                ).replace("\r", "").split("\n")
+                self.logger.debug("Received Data: {data}")
```
I guess this is meant as an f-string. Please convert to lazy % formatting.

JoshuaWatt replied:
Per previous discussion, `%` is not lazt

Bastian-Krause commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,178 @@
 
 def gen_marker():
-    return ''.join(random.choice(string.ascii_uppercase) for i in range(10))
+    return "".join(random.choice(string.ascii_uppercase) for i in range(10))
```
No need to change this line.

Bastian-Krause commented on labgrid/util/timeout.py:
```diff
@@ -23,3 +24,43 @@ def remaining(self):
+class OptionalTimeout:
+    """
+    Represents an optional timeout. If the timeout value is None, the timer
+    will never expire. If timeout negative and default is default is not None,
+    the default timer value will be used
```
This sentence needs revisiting.

Bastian-Krause reviewed (commented):
I've tested this briefly now.

Issues I stumbled upon:
- `BareboxDriver` now times out on `run()`
- running multiple processes on `BareboxDriver`/`ShellDriver` is not forbidden, but runs in timeouts
- calling `stop()` on `ConsoleMarkerProcess` returned by `ShellDriver.popen()` leads to a timeout (e.g. stopping a "sleep 100") - the implementation looks incomplete, I don't see the prompt being detected on `stop()`

Open questions:
- why is `SSHDriver.run()` not using `popen()`?
- I do not really understand the difference between `read_nonblocking()`/`read()`

Styling:
- we use f-strings where appropriate since #838, please do the same

Note: this is no comprehensive review yet, just a collection of things I saw while testing with a QEMU and a physical target.

JoshuaWatt commented:
> I've tested this briefly now.
> 
> Issues I stumbled upon:
> 
>     * `BareboxDriver` now times out on `run()`

I don't have any way of testing barebox, can you provide more information so I can fix it?

> 
>     * running multiple processes on `BareboxDriver`/`ShellDriver` is not forbidden, but runs in timeouts

Do we want to explicitly disallow that? I'm fine with that if that's what everyone wants to do.

> 
>     * calling `stop()` on `ConsoleMarkerProcess` returned by `ShellDriver.popen()` leads to a timeout (e.g. stopping a "sleep 100") - the implementation looks incomplete, I don't see the prompt being detected on `stop()`
> 
> 
> Open questions:
> 
>     * why is `SSHDriver.run()` not using `popen()`?

It would be nice, but I can't do that without breaking the SSH driver because of the stderr_merge option, which is not possible to implement with popen()

> 
>     * I do not really understand the difference between `read_nonblocking()`/`read()`
> 

`read_nonblocking` allows a `CommandProcessProcotol` to be directly used as the backend for a `popen.spawn` object, since the latter requires the the `read_nonblocking` function to be implemented. `read` is the API for normal folks :)

> 
> Styling:
> 
>     * we use f-strings where appropriate since #838, please do the same

I tried to use f-strings everywhere.... IHMO it's confusing to developers to require different string formatting for different uses (unless there is some really compelling performance reason to do otherwise).

> 
> 
> Note: this is no comprehensive review yet, just a collection of things I saw while testing with a QEMU and a physical target.

Bastian-Krause commented:
> > * `BareboxDriver` now times out on `run()`
> > ```
> 
> I don't have any way of testing barebox, can you provide more information so I can fix it?

Here is my demo script:
```python
#!/usr/bin/env python3

import labgrid

labgrid.StepReporter()

env = labgrid.Environment("beaglebone.yaml")
target = env.get_target()

barebox = target.get_driver("BareboxDriver")
target.activate(barebox)

barebox.run("echo hi")
```

This results in:
[labgrid-barebox-run.log](https://github.com/labgrid-project/labgrid/files/7247463/labgrid-barebox-run.log)

I have not looked into it yet.

JoshuaWatt commented:
> How would you run a command that takes longer than 30 seconds?
> 
> ```
> def test_async_output(strategy):
>     strategy.transition("shell")
>     with strategy.shell.popen("sleep 40") as p:
>         pass
> ```
> 
> This ends up with a `TIMEOUT` exception.
> 
> Another thing: how would you read the output of a long-running command (e.g. `find /`) asynchronously? Could you provide an example? I tried something with `while p.poll() is None`, but the `poll()` method blocks for a second if the process is still alive, so that feels wrong.

`poll` is for checking if the process is still alive; to read output use `p.read()` or `p.expect()`

JoshuaWatt commented:
> > How would you run a command that takes longer than 30 seconds?
> > ```
> > def test_async_output(strategy):
> >     strategy.transition("shell")
> >     with strategy.shell.popen("sleep 40") as p:
> >         pass
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This ends up with a `TIMEOUT` exception.
> 
> Ah, this has nothing to do with the long-running command. The context manager hits CTRL-c on close, which prevents the second marker from being echoed (on the shell, not in barebox though). That's why we run into the timeout there.

Does something need to change? It seems like it's doing what you are asking for (launching the process, then immediately terminating it). If you want to wait for it to exit normally, use `p.poll()` in a loop

Bastian-Krause commented:
> > How would you run a command that takes longer than 30 seconds?
> > ```
> > def test_async_output(strategy):
> >     strategy.transition("shell")
> >     with strategy.shell.popen("sleep 40") as p:
> >         pass
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This ends up with a `TIMEOUT` exception.
> > Another thing: how would you read the output of a long-running command (e.g. `find /`) asynchronously? Could you provide an example? I tried something with `while p.poll() is None`, but the `poll()` method blocks for a second if the process is still alive, so that feels wrong.
> 
> `poll` is for checking if the process is still alive; to read output use `p.read()` or `p.expect()`

Okay, will try that.

> > > How would you run a command that takes longer than 30 seconds?
> > > ```
> > > def test_async_output(strategy):
> > >     strategy.transition("shell")
> > >     with strategy.shell.popen("sleep 40") as p:
> > >         pass
> > >   
> > > This ends up with a `TIMEOUT` exception.
> > 
> > 
> > Ah, this has nothing to do with the long-running command. The context manager hits CTRL-c on close, which prevents the second marker from being echoed (on the shell, not in barebox though). That's why we run into the timeout there.
> 
> Does something need to change? It seems like it's doing what you are asking for (launching the process, then immediately terminating it). If you want to wait for it to exit normally, use `p.poll()` in a loop, or `p.wait()`

I think the behavior between barebox and shell should be consistent. I would not expect to get a `TIMEOUT` exception if I do not wait for the process to finish. Do you agree?

JoshuaWatt commented:
> > > How would you run a command that takes longer than 30 seconds?
> > > ```
> > > def test_async_output(strategy):
> > >     strategy.transition("shell")
> > >     with strategy.shell.popen("sleep 40") as p:
> > >         pass
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This ends up with a `TIMEOUT` exception.
> > > Another thing: how would you read the output of a long-running command (e.g. `find /`) asynchronously? Could you provide an example? I tried something with `while p.poll() is None`, but the `poll()` method blocks for a second if the process is still alive, so that feels wrong.
> > 
> > 
> > `poll` is for checking if the process is still alive; to read output use `p.read()` or `p.expect()`
> 
> Okay, will try that.
> 
> > > > How would you run a command that takes longer than 30 seconds?
> > > > ```
> > > > def test_async_output(strategy):
> > > >     strategy.transition("shell")
> > > >     with strategy.shell.popen("sleep 40") as p:
> > > >         pass
> > > >   
> > > > This ends up with a `TIMEOUT` exception.
> > > > ```
> > > 
> > > 
> > > Ah, this has nothing to do with the long-running command. The context manager hits CTRL-c on close, which prevents the second marker from being echoed (on the shell, not in barebox though). That's why we run into the timeout there.
> > 
> > 
> > Does something need to change? It seems like it's doing what you are asking for (launching the process, then immediately terminating it). If you want to wait for it to exit normally, use `p.poll()` in a loop, or `p.wait()`
> 
> I think the behavior between barebox and shell should be consistent. I would not expect to get a `TIMEOUT` exception if I do not wait for the process to finish. Do you agree?

Ya, that makes sense

JoshuaWatt commented:
@Bastian-Krause Ok, I tracked this down to a behavior difference is bash (which I was using) vs. busybox. I've corrected it and now it should work the same for both shells

JoshuaWatt commented:
For those interested, the difference is:

bash:
```
raspberrypi4-64:~# echo -n 'ABC'; sleep 10; echo 'ABC' $?; 
ABC^C
ABC 130
raspberrypi4-64:~#
```

busybox:
```
raspberrypi4-64:~# echo -n 'ABC'; sleep 10; echo 'ABC' $?;
ABC^C
raspberrypi4-64:~#
```

Note that interrupting the command on busybox prevents all subsequent commands (the `echo` with the exit code) from running, where as they still run under bash.

jluebbe commented on labgrid/protocol/commandprotocol.py:
```diff
@@ -1,6 +1,87 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def read_nonblocking(self, size=1, timeout=-1):
+        """
+        This reads at most size characters from the process, or until EOF or a
+        timeout occurs. If no data is read before EOF or a timeout occurs, then
+        pexpect.EOF or pexpect.TIMEOUT are raised respectively.
+
+        A timeout of None means to wait forever for either data or EOF
```
Why does a 'nonblocking' read support waiting for a timeout?

jluebbe replied:
Ah, this matches https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.read_nonblocking

jluebbe commented on labgrid/driver/ubootdriver.py:
```diff
@@ -69,26 +69,25 @@ def on_deactivate(self):
-            exitcode = int(data[-1])
-            del data[-1]
-            return (data, [], exitcode)
+        cmp_command = f"""echo '{marker[:4]}''{marker[4:]}'\\\\c; {cmd}; echo '{marker[:4]}''{marker[4:]}' $?;"""  # pylint: disable=line-too-long
```
What does the `\\\\c` do?

JoshuaWatt replied:
`\c` is the u-boot method for suppressing new-lines in an echo

JoshuaWatt replied:
Yes

jluebbe replied:
Perhaps a comment regarding the `\\\\c` would be useful, though. :)

jluebbe commented on labgrid/driver/sshdriver.py:
```diff
@@ -211,6 +281,27 @@ def _run(self, cmd, codec="utf-8", decodeerrors="strict", timeout=None): # pylin
+        if not self._check_keepalive():
+            raise ExecutionError("Keepalive no longer running")
+
+        complete_cmd = ["ssh", "-o", "LogLevel=QUIET", "-x", *self.ssh_prefix,
+                        "-p", str(self.networkservice.port), "-l", self.networkservice.username,
+                        self.networkservice.address, "-T", "--", '/bin/sh -c {}'.format(shlex.quote(cmd)),
+                        ]
+        self.logger.debug("Sending command: %s", complete_cmd)
```
In the python standard library, popen doesn't use a shell by default. What is the motivation to do this here? Having the same interface as the `ShellDriver`?

JoshuaWatt replied:
Yes, it should behave the same as ShellDriver; I think specifically you need to use a shell for redirection to work

jluebbe commented on labgrid/util/marker.py:
```diff
@@ -1,6 +1,187 @@
+        if not self._alive:
+            return self.exitcode
+
+        index, _, match, _ = self._console.expect([self._eof_re, TIMEOUT], timeout=1)
```
This looks like it could drop output from the process. Is that intentional?

JoshuaWatt replied:
I don't think it can because the EOF is anchored to the beginning of the string, so unless the output has all been consumed, the process will still be "alive" (see the warning in `wait`)

jremmet commented on labgrid/protocol/commandprotocol.py:
```diff
@@ -1,6 +1,87 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def stop(self):
```
nice work :)
I tested it with my setup and run into a issue while running:
```
def test_shell_interrupt(shell_command):
    with shell_command.popen("yes") as p:
        l = p.read(1024).decode("utf-8").splitlines()
        assert l == ["y"] * len(l)

    shell_command.run_check("true")
```

on the console log I see:
```
root@ksp0627-imx6-1:~# MARKER='XVUS''LJFJHA' run yes
XVUSLJFJHAy
y
y


y
y
^C
root@ksp0627-imx6-1:~#
```

put the test runs into a 60sec  timout
```

        SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!YXLELHEROR)(?!YXLELHERO$)(?!YXLELHER$)(?!YXLELHE$)(?!YXLELH$)(?!YXLEL$)(?!YXLE$)(?!YXL$)(?!YX$)(?!Y$).', re.DOTALL), re.compile(b'^YXLELHEROR\\s+(\\d+)\\s+.*root@[\\w-]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}




       SerialDriver.expect state='stop' result=(2, b'', <class 'pexpect.exceptions.TIMEOUT'>, <class 'pexpect.exceptions.TIMEOUT'>) duration='60.007'
     ConsoleMarkerProcess.expect state='stop' result=(1, b'\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\ny\r\n^C\r\n', <re.Match object; span=(711, 733), match=b'root@ksp0627-imx6-1:~#'>, b'root@ksp0627-imx6-1:~#') duration='60.087'
   ConsoleMarkerProcess.stop state='stop' duration='60.088'
```

Seams `read_nonblocking` not give the "promt" only case without a timout for busybox shells

JoshuaWatt replied:
@jremmet Hmm, that suspiciously looks like the problem I fixed yesterday with busybox.... can you please make sure verify you have the latest commits I pushed? If you do please let me know and I'll look more; specifically, there should a additional regex that matches the prompt in the `SerialDriver.expect` (for a total of 3 regex it's looking for)

jremmet replied:
Thanks for the fast response. I'm on 83c64e9 JoshuaWatt/labgrid.git@command-background. The timed out expext seams to come from `ConsoleMarkerProcess` `read_nonblocking`. There I found no `_prompt_re`.

JoshuaWatt replied:
Huh, I'm not sure what is going on here; I've been testing with busybox and can't seem to reproduce what's going on here. The weird part is that `stop` is waiting 60 seconds but then returning that if found the prompt anyway

Can you do a little more digging and provides some more logs?

jremmet replied:
After updating to 58d847d to timeout is gone :+1:

jremmet reviewed (commented):
Nice work, found a issue on a busybox system

JoshuaWatt committed:
util: Add readmixin class

Adds a read mix in class that implements read helper functions to deal
with reading from sources that will return only buffered data, instead
of trying harder to read up to the user requested size.

These functions are modeled after the functions provided by Rust

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt commented:
Ping

JoshuaWatt commented:
Ping

JoshuaWatt commented:
Ping

JoshuaWatt commented:
Ping

Emantor commented:
Hi,

I deployed this internally and let it run over our internal test suites and didn't find an error related to this PR.
Thats great, thanks a lot for the work!
The issue I see at the moment is that the verbose pytest output becomes unreadable:
```
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'y'>, b'y')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b't'>, b't')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'e'>, b'e')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'm'>, b'm')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'/'>, b'/')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'y'>, b'y')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b't'>, b't')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'e'>, b'e')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'm'>, b'm')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'd'>, b'd')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'-'>, b'-')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'a'>, b'a')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'u'>, b'u')
10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
```
Since we match on a single byte a lot of the steps don't contain any meaningful information. I am not sure how to solve this, one way would be to create a helper which does not contain the step decorator to supress this output. Note that `expect()` is still called by drivers and I wouldn't want to get rid of this as is.

JoshuaWatt commented:
> Hi,
> 
> I deployed this internally and let it run over our internal test suites and didn't find an error related to this PR. Thats great, thanks a lot for the work! The issue I see at the moment is that the verbose pytest output becomes unreadable:
> 
> ```
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'y'>, b'y')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b't'>, b't')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'e'>, b'e')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'm'>, b'm')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'/'>, b'/')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'y'>, b'y')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b's'>, b's')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b't'>, b't')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'e'>, b'e')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'm'>, b'm')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'd'>, b'd')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'-'>, b'-')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'a'>, b'a')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> 10:13:34      SerialDriver.expect state='stop' result=(0, b'', <re.Match object; span=(0, 1), match=b'u'>, b'u')
> 10:13:34      SerialDriver.expect state='start' args={'pattern': [re.compile(b'^(?!WASPAEMYUO)(?!WASPAEMYU$)(?!WASPAEMY$)(?!WASPAEM$)(?!WASPAE$)(?!WASPA$)(?!WASP$)(?!WAS$)(?!WA$)(?!W$).', re.DOTALL), re.compile(b'^WASPAEMYUO\\s+(\\d+)\\s+.*root@[^:]+:[^ ]+'), <class 'pexpect.exceptions.TIMEOUT'>]}
> ```
> 
> Since we match on a single byte a lot of the steps don't contain any meaningful information. I am not sure how to solve this, one way would be to create a helper which does not contain the step decorator to supress this output. Note that `expect()` is still called by drivers and I wouldn't want to get rid of this as is.

Yes, I noticed that; do you have any suggestions for how to make it more reasonable?

Bastian-Krause commented on labgrid/util/readmixin.py:
```diff
@@ -0,0 +1,77 @@
+from .timeout import Timeout
+from pexpect import EOF, TIMEOUT
```
```suggestion
from pexpect import EOF, TIMEOUT

from .timeout import Timeout
```
nitpick: wrong import order

JoshuaWatt commented:
> Yes, I noticed that; do you have any suggestions for how to make it more reasonable?

Hah, you did give a suggestion! My fault for not reading thoroughly. I'll look into doing that.

JoshuaWatt committed:
sshdriver: Disable input echo

Disables echo in the shell when using start_process() with the ssh
driver. This prevents duplicate output so it behaves as the other
drivers.

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JoshuaWatt committed:
sshdriver: Run start_process() commands in a PTY

Instructs ssh to open a pseudo terminal when running start_process()
commands.  This allows signals to propagate to the remote process
(instead of being intercepted by the ssh client process), emulating the
start_process() signal behavior on other drivers.

Signed-off-by: Joshua Watt <Joshua.Watt@garmin.com>

JPEWdev commented:
@Bastian-Krause I think this is ready, would you mind looking again?

Bastian-Krause commented:
Thanks for keeping at it! I'll try to have a look at this and run our network test suite against this.

JoshuaWatt commented on labgrid/util/marker.py:
```diff
@@ -1,9 +1,174 @@
+    """labgrid Wrapper of the pexpect module.
+
+    This class provides pexpect functionality for the ConsoleProtocol classes.
+    driver: ConsoleProtocol object to be passed in
```
Done

JoshuaWatt committed:
Merge branch 'master' into command-background

JoshuaWatt commented:
Would someone mind looking at this change again? Thanks

Emantor reviewed (approved):
I gave this a once over and everything seems to be addressed. We just had a release, so its a good time to get this in for additional testing.

Emantor commented:
We took a deeper look with @jluebbe, @Bastian-Krause and me and taking the state of our serial stack into account, we are not really comfortable to add so much additional complexity on top of the rather fragile expect matching. Nevertheless, the changes for the SSHDriver are much less invasive and more straightforward. What do you think about splitting these out and getting them merged?

While reviewing we also wondered whether the background functionality for the bootloader drivers is really necessary, do you have examples for usage or were those implemented for CommandProtocol completeness?

JoshuaWatt commented:
> We took a deeper look with @jluebbe, @Bastian-Krause and me and taking the state of our serial stack into account, we are not really comfortable to add so much additional complexity on top of the rather fragile expect matching. Nevertheless, the changes for the SSHDriver are much less invasive and more straightforward. What do you think about splitting these out and getting them merged?
> 
> While reviewing we also wondered whether the background functionality for the bootloader drivers is really necessary, do you have examples for usage or were those implemented for CommandProtocol completeness?

Yes, it was mostly done for completeness. I don't beleive we would need serial async execution. Do you want the serial protocols to raise a `NotImplementedError` for `CommandProtocol.start_process()` or would you prefer `start_process()` be part of a new protocol?

Emantor commented:
We should just implement it on top of the SSHDriver and not use a protocol at all. If we ever find a second user we can still extend/create a protocol.

Bastian-Krause reviewed (commented):
Since `SSHDriver.start_process()` merges stdout and stderr, I'm wondering if it should raise `NotImplementedError` if its `stderr_merge` attribute is False?

JoshHollis251 commented:
> Since `SSHDriver.start_process()` merges stdout and stderr, I'm wondering if it should raise `NotImplementedError` if its `stderr_merge` attribute is False?

If we require stderr_merge=false and the ssh_driver instance were to perform a .run() following a start_process() operation it would have the side effect of maybe unintentionally merging those fields unless it is set back to the default. Is that acceptable behavior, and should we include a unit test for the not_implemented check?

JoshuaWatt commented:
FYI: @JoshHollis251 is helping me finish this PR :)

Bastian-Krause commented:
> > Since `SSHDriver.start_process()` merges stdout and stderr, I'm wondering if it should raise `NotImplementedError` if its `stderr_merge` attribute is False?
> 
> If we require stderr_merge=false and the ssh_driver instance were to perform a .run() following a start_process() operation it would have the side effect of maybe unintentionally merging those fields unless it is set back to the default. Is that acceptable behavior, and should we include a unit test for the not_implemented check?

After discussing this with @Emantor: we're not happy with the `stderr_merge` attribute in general, it should have been a kwarg in `run()`/`run_check()`/`wait_for()` from the start. So if we ever get rid of it in the long run, we should not rely on it here.
Also: Since separating stdout from stderr is worth keeping and the default, we should not artificially enforce `stderr_merge=False` just so `start_process()` can be used. So: let's not check that attribute and let's not raise `NotImplementedError`. Let us instead add a docstring documenting that stdout and stderr are merged in `start_process()`.

nick-potenski commented on tests/test_sshdriver.py:
```diff
@@ -218,3 +219,92 @@ def test_unix_socket_forward(ssh_localhost, tmpdir):
+def test_start_process_read_to_end(ssh_localhost):
+    with ssh_localhost.start_process("echo Hello World") as p:
+        p.read_to_end() == b"Hello World"
+        p.read_to_end() == b""
```
Missing assert on both of these lines?

nick-potenski commented on labgrid/util/readmixin.py:
```diff
@@ -0,0 +1,81 @@
+
+class ReadMixIn:
+    """
+    This class make it more convenient to deal with reading from devices. A
```
```suggestion
    This class makes it more convenient to deal with reading from devices. A
```
